src/crypto/crypto_context.cc

```diff
@@ -222,6 +227,272 @@ unsigned long LoadCertsFromFile(  // NOLINT(runtime/int)
+      return true;
+    }
+  }
+
+  CFIndex trustSettingsCount = CFArrayGetCount(trustSettings);
+
+  for (CFIndex i = 0; i < trustSettingsCount ; ++i) {
+    CFDictionaryRef trustDict = (CFDictionaryRef) CFArrayGetValueAtIndex(
+        trustSettings, i);
+
+    enum TrustStatus trust = IsTrustDictionaryTrustedForPolicy(trustDict);
+
+    if (trust == DISTRUSTED) {
+      return false;
+    } else if (trust == TRUSTED) {
+      return true;
+    }
+  }
+  return false;
+}
+
+bool IsCertificateTrustValid(SecCertificateRef ref) {
+  SecTrustRef secTrust = nullptr;
+  CFMutableArrayRef subjCerts = CFArrayCreateMutable(
+      nullptr, 1, &kCFTypeArrayCallBacks);
+  CFArraySetValueAtIndex(subjCerts, 0, ref);
+
+  SecPolicyRef policy = SecPolicyCreateBasicX509();
+  OSStatus ortn = SecTrustCreateWithCertificates(subjCerts, policy, &secTrust);
+  bool result = false;
+  if (ortn) {
+    /* should never happen */
+    goto errOut;
+  }
+
+  result = SecTrustEvaluateWithError(secTrust, nullptr);
+  errOut:
+     if (policy) {
+       CFRelease(policy);
+     }
+  if (secTrust) {
+    CFRelease(secTrust);
+  }
+  if (subjCerts) {
+    CFRelease(subjCerts);
+  }
+  return result;
+}
+
+bool IsCertificateTrustedForPolicy(X509* cert, SecCertificateRef ref) {
+  OSStatus err;
+
+  for (const auto& trust_domain :
+       {kSecTrustSettingsDomainUser, kSecTrustSettingsDomainAdmin}) {
+    CFArrayRef trustSettings;
+    err = SecTrustSettingsCopyTrustSettings(ref, trust_domain, &trustSettings);
+
+    bool isSelfSigned = IsSelfSigned(cert);
+
+    if (err == errSecSuccess && trustSettings != nullptr) {
+      return IsTrustSettingsTrustedForPolicy(trustSettings, isSelfSigned);
+    }
+
+    // An empty trust settings array isn’t the same as no trust settings,
+    // where the trustSettings parameter returns NULL.
+    // No trust-settings array means
+    // “this certificate must be verifiable using a known trusted certificate”.
+    if (trustSettings == nullptr) {
+      return IsCertificateTrustValid(ref);
+    }
+  }
+  return false;
+}
+
+void ReadMacOSKeychainCertificates(
+    std::vector<std::string>* system_root_certificates) {
+  CFTypeRef searchKeys[] = { kSecClass, kSecMatchLimit, kSecReturnRef };
+  CFTypeRef searchValues[] = {
+      kSecClassCertificate, kSecMatchLimitAll, kCFBooleanTrue };
+  CFDictionaryRef search = CFDictionaryCreate(
+      kCFAllocatorDefault, searchKeys, searchValues, 3,
+      &kCFTypeDictionaryKeyCallBacks, &kCFTypeDictionaryValueCallBacks);
+
+  CFArrayRef currAnchors = nullptr;
+  OSStatus ortn = SecItemCopyMatching(
+      search,
+      reinterpret_cast<CFTypeRef *>(&currAnchors));
+
+  if (ortn) {
+      fprintf(stderr, "ERROR: SecItemCopyMatching failed %d\n", ortn);
+  }
+
+  CFIndex count = CFArrayGetCount(currAnchors);
+
+  std::vector<X509*> system_root_certificates_X509;
+  for (int i = 0; i < count ; ++i) {
+    SecCertificateRef certRef = (SecCertificateRef) CFArrayGetValueAtIndex(
+        currAnchors, i);
```

Comment:
This should use `reinterpret_cast`

Reply:
This fails to build with:

```
../../src/crypto/crypto_context.cc:447:33: error: reinterpret_cast from 'const void *' to 'SecCertificateRef' (aka '__SecCertificate *') casts away qualifiers
  447 |     SecCertificateRef certRef = reinterpret_cast<SecCertificateRef>(CFArrayGetValueAtIndex(
```

The linter hasn't asked me to change to it and it did in most of the other places.

Reply:
Then it's const_cast + reinterpret_cast here, I guess :)

Reply:
```c++
SecCertificateRef certRef = const_cast<SecCertificateRef>(reinterpret_cast<const SecCertificateRef>(CFArrayGetValueAtIndex(
        currAnchors, i)));
```

fails with:
> error: reinterpret_cast from 'const void *' to 'SecCertificateRef' (aka '__SecCertificate *') casts away qualifiers

```c++
SecCertificateRef certRef = const_cast<SecCertificateRef>(CFArrayGetValueAtIndex(currAnchors, i));
```
fails with:
> error: const_cast from 'const void *' to 'SecCertificateRef' (aka '__SecCertificate *') is not allowed

I've tried quite a few options and I haven't managed to get something to compile here

Reply:
I checked chromium's source and found out how to do it:

```
    SecCertificateRef certRef = reinterpret_cast<SecCertificateRef>(
        const_cast<void*>(CFArrayGetValueAtIndex(currAnchors, i)));
```